### PR TITLE
Reverse split

### PR DESCRIPTION
**Pull request by gsquire**
This is an attempt at #19 

I don't know where a good place for unit tests would be but I would like to add some.

### CODE REVIEW

File: library.go [hunk 1]
```diff
@@ -1793,8 +1793,7 @@ func string_split(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Value
 		} else if fnname == "split" {
 			res = splitspace(recv, maxsplit+1)
 		} else { // rsplit
-			// TODO(adonovan): implement.
-			return nil, fmt.Errorf("rsplit(None, %d): maxsplit > 0 not yet supported", maxsplit)
+			res = rsplit(recv, maxsplit)
```
Comment (alandonovan) [inline library.go]: > I don't know where a good place for unit tests would be but I would like to add some.

The dynamic tests live in the testdata/*.sky files.  In testdata/string.sky, you'll find the existing tests for rsplit.  You can put addition checks for rsplit there.

File: library.go [hunk 2]
```diff
@@ -1827,6 +1826,49 @@ func string_split(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Value
 	return NewList(list), nil
 }
 
+func reverse(s string, size int) string {
```
Comment (alandonovan) [inline library.go]: This doesn't warrant its own function.  Just inline the sole call.

File: library.go [hunk 3]
```diff
@@ -1827,6 +1826,49 @@ func string_split(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Value
...
+
+	return string(rev)
+}
+
+func rsplit(s string, max int) []string {
+	size := utf8.RuneCountInString(s)
+	splitCount := 0   // Our current number of splits made.
+	currIndex := size // A backward working index.
+	lastStop := size  // After we make a split, advance this back.
+	parts := make([]string, 0, max)
```
Comment (alandonovan) [inline library.go]: Could you add a test for max < 0?

File: library.go [hunk 4]
```diff
@@ -1827,6 +1826,49 @@ func string_split(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Value
...
+		rev[i], rev[size-1-i] = rev[size-1-i], rev[i]
+	}
+
+	return string(rev)
+}
+
+func rsplit(s string, max int) []string {
+	size := utf8.RuneCountInString(s)
+	splitCount := 0   // Our current number of splits made.
+	currIndex := size // A backward working index.
```
Comment (alandonovan) [inline library.go]: Call this i, and move it closer to the loop.  No need for the comment.

File: library.go [hunk 5]
```diff
@@ -1827,6 +1826,49 @@ func string_split(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Value
...
+
+	for _, point := range s {
+		rev = append(rev, point)
+	}
+
+	for i := 0; i < size/2; i++ {
+		rev[i], rev[size-1-i] = rev[size-1-i], rev[i]
+	}
+
+	return string(rev)
```
Comment (alandonovan) [inline library.go]: This allocates a copy as large as the input.  How hard would it be to avoid materializing the reversed list?

It should be possible to discover the index of each new code point by working back at most 4 bytes from the previous index.

File: library.go [hunk 6]
```diff
@@ -1827,6 +1826,49 @@ func string_split(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Value
...
+func rsplit(s string, max int) []string {
+	size := utf8.RuneCountInString(s)
+	splitCount := 0   // Our current number of splits made.
+	currIndex := size // A backward working index.
+	lastStop := size  // After we make a split, advance this back.
+	parts := make([]string, 0, max)
+
+	reversed := reverse(s, size)
+
+	for _, p := range reversed {
```
Comment (alandonovan) [inline library.go]: This loop does UTF-8 decoding a second time.  It would be more efficient (and probably fewer lines of code) to work backwards, identifying the start of each UTF-8 encoded code point, in a single pass.

File: library.go [hunk 7]
```diff
@@ -1827,6 +1826,49 @@ func string_split(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Value
...
+
+			splitCount++
+			lastStop = currIndex - 1
+		}
+		currIndex--
+	}
+
+	res := []string{s[0:lastStop]}
+	for i := len(parts) - 1; i >= 0; i-- {
+		res = append(res, parts[i])
```
Comment (alandonovan) [inline library.go]: This allocates a new slice that is essentially the reverse of parts.  It would be more efficient to append the final string to parts, then reverse parts in place (using the logic at lines 1836-38).

**Comment by alandonovan** [comment]
Hi Garrett, many thanks for contributing this fix, and sorry it took me so long to review it.  The logic looks sound, but I have noted some suggestions for making it much faster---while hopefully not more complicated---by allocating less memory.

cheers
alan

**Comment by gsquire** [comment]
Thanks for the review, I know I made a few more allocations than would be ideal. I will write some tests and follow the rest of your comments to improve this diff.

File: library.go (hunk 5 above)
Comment (gsquire) [reply library.go]: Since I will just move this into the function, I can iterate through the rune slice which should be safe to do. Please correct me if I am wrong.

File: library.go (hunk 3 above)
Comment (gsquire) [reply library.go]: Heh, I forgot you could pass -1 to `rsplit` to signify no limits. I will just remove the capacity in this `make` statement.

File: library.go (hunk 6 above)
Comment (gsquire) [reply library.go]: I am just going to iterate through the slice as noted above.

File: library.go (hunk 7 above)
Comment (gsquire) [reply library.go]: Good idea, I will change this.

File: library.go (hunk 3 above)
Comment (alandonovan) [reply library.go]: I just realized that the caller handles the zero and negative cases, thus your function is guaranteed  `max > 0`.  You can ignore my comment.

**Comment by gsquire** [comment]
I only had to make a minor tweak for your implementation to work. I will think of some good test cases now.

Let me know if you would like me to squash my commits.

File: library.go [hunk 8]
```diff
@@ -1793,8 +1793,7 @@ func string_split(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Value
 		} else if fnname == "split" {
 			res = splitspace(recv, maxsplit+1)
```
Comment (alandonovan) [inline library.go]: While you're here, for symmetry, could you remove the +1 on this line and in the expression len(res)+1 == max inside splitspace too?

File: library.go [hunk 9]
```diff
@@ -1827,6 +1826,36 @@ func string_split(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Value
 	return NewList(list), nil
 }
 
+func rsplitspace(s string, max int) []string {
```
Comment (alandonovan) [inline library.go]: Add comment:

// Precondition: max > 0.

File: library.go [hunk 10]
```diff
@@ -1827,6 +1826,36 @@ func string_split(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Value
...
+
+	resLen := len(res)
+	for i := 0; i < resLen/2; i++ {
+		res[i], res[resLen-1-i] = res[resLen-1-i], res[i]
+	}
+
+	return res
+}
+
 func splitspace(s string, max int) []string {
```
Comment (alandonovan) [inline library.go]: Add comment:
// Precondition: max > 0.

File: library.go [hunk 11]
```diff
@@ -1827,6 +1826,36 @@ func string_split(fnname string, recv_ Value, args Tuple, kwargs []Tuple) (Value
...
 
+func rsplitspace(s string, max int) []string {
+	res := make([]string, 0, max+1)
+	end := -1 // index of field end, or -1 in a region of spaces.
+	for i := len(s); i > 0; {
+		r, sz := utf8.DecodeLastRuneInString(s[:i])
+		if unicode.IsSpace(r) {
+			if end >= 0 {
+				if len(res) == max {
+					break // let this field run to the start.
```
Comment (alandonovan) [inline library.go]: (FWIW, for inline comments I don't typically worry about capitals/period even when they are sentences, as in this case.)

File: testdata/string.sky [hunk 12]
```diff
@@ -189,10 +189,16 @@ assert.eq("a.b.c.d".rsplit(".", 2), ["a.b", "c", "d"])
...
+assert.eq(" a bc\n  def \t  ghi".rsplit(None, 0), [" a bc\n  def \t  ghi"])
 assert.eq(" a bc\n  def \t  ghi".split(None, 1), ["a", "bc\n  def \t  ghi"])
+assert.eq(" a bc\n  def \t  ghi".rsplit(None, 1), [" a bc\n  def", "ghi"])
 assert.eq(" a bc\n  def \t  ghi".split(None, 2), ["a", "bc", "def \t  ghi"])
+assert.eq(" a bc\n  def \t  ghi".rsplit(None, 2), [" a bc", "def", "ghi"])
 assert.eq(" a bc\n  def \t  ghi".split(None, 3), ["a", "bc", "def", "ghi"])
+assert.eq(" a bc\n  def \t  ghi".rsplit(None, 3), [" a", "bc", "def", "ghi"])
 assert.eq(" a bc\n  def \t  ghi".split(None, 4), ["a", "bc", "def", "ghi"])
+assert.eq(" a bc\n  def \t  ghi".rsplit(None, 4), ["a", "bc", "def", "ghi"])
+assert.eq(" a bc\n  def \t  ghi".rsplit(None, 5), ["a", "bc", "def", "ghi"])
```
Comment (alandonovan) [inline testdata/string.sky:201]: Can you add a test case with trailing space after ghi, for both rsplit and rsplit, both zero and nonzero?

**Review by alandonovan — COMMENTED** [review]
Glad it worked!  Thanks, this is nearly done.

File: library.go (hunk 8 above)
Comment (gsquire) [reply library.go]: Sure, no problem.

File: library.go (hunk 11 above)
Comment (gsquire) [reply library.go]: Ok, I will remove it.

File: testdata/string.sky (hunk 12 above)
Comment (gsquire) [reply testdata/string.sky:201]: Sure thing.

**Review by alandonovan — APPROVED** [review]
Many thanks!

**Comment by gsquire** [comment]
I saw that the tests were broken after I merged, but it looks like you made changes elsewhere?

**Comment by alandonovan** [comment]
Yes, the test I suggested you add, of sep=None maxsplit=0, revealed a problem, fixed in https://github.com/google/skylark/pull/35. Essentially, sep=None is supposed to use a different algorithm from sep="...".  The actual code change was very small.
